File: frontend/projects/irrtirer/src/app/shared/mosaic-generator/orderer/section-populator.ts
import { BluredImageObject } from '../color-compatibility/blured-image-object.model';
import { SeparatingAxisTheorem } from '../helpers/polygon-helpers/separating-axis-theorem';
import { Tile } from '../models/tile';
import { TileTransform } from '../models/tile-transform.model';
import { PopulationParams } from '../sectors/generation-params/population-params.interface';
import { SectionModel } from '../sectors/section.model';
import { PossibleLayout } from './section-layout/possible-layout';
import { SectionLayoutPreparer } from './section-layout/section-layout-preparer';
import { SpecimenSelection } from './section-layout/specimen-selection';

export class SectionPopulator {
    private readonly populationParams: PopulationParams;
    private readonly specimenSelection: SpecimenSelection;
    private readonly sectionLayoutPreparer: SectionLayoutPreparer;

    private readonly K: number;

    constructor(pixelSource: BluredImageObject, section: SectionModel) {
        this.populationParams = section.parent.populationParams;
        this.K = Math.max(2, this.populationParams.populationSize / 4);
        this.specimenSelection = new SpecimenSelection();
        this.sectionLayoutPreparer = new SectionLayoutPreparer(pixelSource, section, this.getSectionNeighbourTiles(section));
    }

    public searchForBestSectionLayout(avalibleTiles: Tile[]): PossibleLayout {
        const initialPopulation: PossibleLayout[] = this.prepareRandomPopulation(avalibleTiles);
        const population: PossibleLayout[] = new Array(this.populationParams.populationSize);
        const tempArray: PossibleLayout[] = new Array(this.populationParams.populationSize);

        this.selectBestPopulation(initialPopulation, population, this.K);
        for (let i = 0; i < this.populationParams.iterationsCount; ++i) {
            //this.mutatePopulation(population, tempArray);
            //this.selectBestPopulation(tempArray, population, this.K);
        }

        return this.specimenSelection.selectBestInPopulation(population);
    }

    private mutatePopulation(oldPopulation: PossibleLayout[], newPopulation: PossibleLayout[]): void {
        throw new Error();
    }

    private prepareRandomPopulation(avalibleTiles: Tile[]): PossibleLayout[] {
        const initialPopulation: PossibleLayout[] = new Array(this.populationParams.initialPopulationSize);
        //tu byli paralele
        for (let i = 0; i < this.populationParams.initialPopulationSize; ++i) {
            initialPopulation[i] = this.sectionLayoutPreparer.createInitialLayout(avalibleTiles);
        }

        return initialPopulation;
    }

    private selectBestPopulation(oldPopulation: PossibleLayout[], newPopulation: PossibleLayout[], k: number): void {
        newPopulation[0] = this.specimenSelection.selectBestInPopulation(oldPopulation);
        for (let i = 1; i < newPopulation.length; i++) {
            newPopulation[i] = this.specimenSelection.turnamentSelection(oldPopulation, k);
        }
    }

    private getSectionNeighbourTiles(section: SectionModel): TileTransform[] {
        return section.neighbours
            .filter((s) => s.wasGenerated)
            .flatMap((s) => s.directTiles)
            .filter((t) => SeparatingAxisTheorem.getMinDistance(t.getWorldVertices(), section.getWorldVertices()) <= section.parent.tileMaxRadius + section.parent.tileMargin);
    }
}

File: backend/Irrtirer/Irrtirer.Generator/Orderer/SectionPopulator.cs
using Irrtirer.Generator.ColorCompatibility;
using Irrtirer.Generator.Helpers;
using Irrtirer.Generator.Helpers.PolygonHelpers;
using Irrtirer.Generator.Models.GenerationParams;
using Irrtirer.Generator.MosaicStructure;
using Irrtirer.Generator.Orderer.SectionLayout;
using Irrtirer.Generator.Tiles;
using Microsoft.Extensions.Logging;
using System;
using System.Linq;
using System.Threading.Tasks;

namespace Irrtirer.Generator.Orderer
{
    internal class SectionPopulator
    {
        private readonly ParallelOptions parallelOptions;
        private readonly PopulationParams populationParams;
        private readonly SpecimenSelection specimenSelection;
        private readonly SectionLayoutPreparer sectionLayoutPreparer;

        private readonly int K;

        private readonly ILogger logger;

        public SectionPopulator(
            ILogger logger,
            RandomFactory randomFactory,
            BlurredImageObject pixelSource,
            ParallelOptions parallelOptions,
            SectionModel section)
        {
            this.logger = logger;
            this.parallelOptions = parallelOptions;
            this.populationParams = section.Parent.PopulationParams;
            K = (int)MathF.Max(2, populationParams.PopulationSize / 4f);
            this.specimenSelection = new SpecimenSelection(randomFactory.GetRandomObject());
            this.sectionLayoutPreparer = new SectionLayoutPreparer(randomFactory, pixelSource, section, GetSectionNeighborsTiles(section));
        }

        public PossibleLayout SearchForBestSectionLayout(Tile[] availableTiles)
        {
            PossibleLayout[] initialPopulation = PrepareRandomPopulation(availableTiles);
            PossibleLayout[] population = new PossibleLayout[populationParams.PopulationSize];
            PossibleLayout[] tempArray = new PossibleLayout[populationParams.PopulationSize];

            SelectBestPopulation(initialPopulation, population, K);
            for (int i = 0; i < populationParams.IterationsCount; ++i)
            {
                MutatePopulation(population, tempArray);
                SelectBestPopulation(tempArray, population, K);
            }

            return specimenSelection.SelectBestInPopulation(population);
        }

        private void MutatePopulation(PossibleLayout[] oldPopulation, PossibleLayout[] newPopulation)
        {
            newPopulation[0] = oldPopulation[0].Clone() as PossibleLayout;
            Parallel.For(1, populationParams.PopulationSize, parallelOptions, (index) =>
            {
                try
                {
                    newPopulation[index] = sectionLayoutPreparer.ModifyLayout(oldPopulation[index].Clone() as PossibleLayout, 3);
                }
                catch (Exception ex)
                {
                    logger.LogError(ex.ToString());
                }
            });
        }

        private PossibleLayout[] PrepareRandomPopulation(Tile[] availableTiles)
        {
            PossibleLayout[] initialPopulation = new PossibleLayout[populationParams.InitialPopulationSize];
            Parallel.For(0, populationParams.InitialPopulationSize, parallelOptions, (index) =>
            {
                try
                {
                    initialPopulation[index] = sectionLayoutPreparer.CreateInitialLayout(availableTiles);
                }
                catch (Exception ex)
                {
                    logger.LogError(ex.ToString());
                }
            });

            return initialPopulation;
        }

        private void SelectBestPopulation(PossibleLayout[] oldPopulation, PossibleLayout[] newPopulation, int k)
        {
            newPopulation[0] = specimenSelection.SelectBestInPopulation(oldPopulation);
            for (int i = 1; i < newPopulation.Length; i++)
            {
                newPopulation[i] = specimenSelection.TournamentSelection(oldPopulation, k);
            }
        }

        private TileTransform[] GetSectionNeighborsTiles(SectionModel section)
        {
            return section.Neighbors
                .Where(s => s.WasGenerated)
                .SelectMany(s => s.DirectTiles)
                .Where(t => SeparatingAxisTheorem.GetMinDistance(t.GetWorldVertices(), section.GetWorldVertices()) <= section.Parent.TileMaxRadius + section.Parent.TileMargin)
                .ToArray();
        }
    }
}
